README.md
# Ring-IP-Whitelist

Ring middleware that allows you to protect routes with an IP whitelist.

## Install

Add the following dependency to your `project.clj`

```
[net.danielcompton/ring-ip-whitelist "0.2.1"]
```

[![Dependencies Status](https://versions.deps.co/danielcompton/ring-ip-whitelist/status.svg)](https://versions.deps.co/danielcompton/ring-ip-whitelist)

## Usage

To IP whitelist routes, wrap them with `wrap-ip-whitelist`. `wrap-ip-whitelist` support IPv4 and IPv6 IP addresses.

```clj
(ns my.cool-ns
  (:require [ring.middleware.ip-whitelist :as ip]))

(def app
  (-> handler
      (ip/wrap-ip-whitelist {:cidrs ["8.8.4.4" "210.50.2.0/30" "::1/128"})))
```

`wrap-ip-whitelist` takes a map of configuration options. Only `:cidrs` is required.

* `:cidrs`: A whitelist of CIDR strings or IP addresses. IP addresses are treated as a CIDR for a single address (e.g. a /32 for an IPv4 address). If the list is inside an atom then the whitelist will be updated when the atom changes.
* `:ip-fn`: An optional function to extract the IP address. Defaults to `:remote-addr`.
* `:error-response`: An optional function that returns a response if the request is not in the whitelist. Defaults to a 403 Not Authorized response.
* `:allow-access?`: An optional function that can be called if IP whitelisting fails but more checking is needed before denying access. Takes a request argument, returns truthy if the request should be allowed, otherwise it will be denied."

CIDR ranges shouldn't be too large. The current implementation generates every IP address in the CIDR range. This is fine for a smallish number of IP addresses in the whitelist, and keeps the implementation simple and fast. The tradeoff is memory space. Don't use this to whitelist the whole world unless you want to run out of memory very shortly afterwards.

#### Passing `:cidrs` as an atom

```clj
(def ip-whitelist (atom ["8.8.4.4"]))
(-> routes
    (ip/wrap-ip-whitelist {:cidrs ip-whitelist}))

(swap! ip-whitelist conj "100.121.5.33/30")

;; Middleware now whitelists these IPs
;; ["8.8.4.4", "100.121.5.32", "100.121.5.33", "100.121.5.34", "100.121.5.35"]
```

#### `:ip-fn`

In almost all cases you should use something like [ring.middleware.proxy-headers/wrap-forwarded-remote-addr](https://ring-clojure.github.io/ring-headers/ring.middleware.proxy-headers.html) to translate proxy headers into `:remote-addr` for you. If for some reason you do need to look at other headers to find the 'real' IP address of a client then you can do it here. `:ip-fn` takes a Ring request and returns an IP address string.

#### `:error-response`

If people fail your IP whitelist, you may want to return a customised 403 error page that matches your sites theme, or a 404 so you don't reveal that the user wasn't authorised to access a protected route. You can do this by passing an `:error-response` function to override the default. `:error-response` takes the failed Ring request, and returns a Ring response.

```clj
(-> routes
    (wrap-ip-whitelist {:cidrs          ["127.0.0.1"]
                        :error-response (constantly
                                          {:status  404
                                           :headers {"Content-Type" "text/html"}
                                           :body    "<h1>Not found</h1>"})}))
```

#### `:allow-access?`

Use `:allow-access?` if you need to do additional checking before denying someone access. For example you may want to allow access from within your internal network to anyone, but outside users have to be authenticated.

```clj
(-> routes
    (ip/wrap-ip-whitelist {:cidrs ["10.0.0.0/28"]
                           :allow-access (fn [req] (check-cookies-to-see-if-authorised req))}))
```


## Security of IP whitelisting

While IP whitelisting increases the difficulty for an attacker, if an attacker controls a router between a whitelisted IP and the server they can bypass your IP whitelisting. Also, if they can get onto the network of one of the whitelisted ranges then it is also game over. Treat IP whitelisting as an additional layer of security, not your only one. Some useful links to read more on IP whitelisting:


* [How secure is IP address filtering?](http://stackoverflow.com/questions/437146/how-secure-is-ip-address-filtering)
* [Is IP whitelist sufficient to protect a server?](http://security.stackexchange.com/questions/51587/is-ip-whitelist-sufficient-to-protect-a-server)
* [The ugly truth about IP whitelisting](https://community.akamai.com/community/cloud-security/blog/2014/11/06/the-ugly-truth-behind-the-practice-of-ip-whitelisting)
* [What security risks does IP spoofing bring?](http://security.stackexchange.com/questions/1009/what-security-risks-does-ip-spoofing-bring)
* [Obscurity is a Valid Security Layer](https://danielmiessler.com/study/security-by-obscurity/)


## Protect some of your routes with IP whitelisting

TODO: Show how you can protect just some of your routes before they are combined.  
TODO: Where to put this in your stack of middleware.

## License

Copyright © 2016 Daniel Compton

This project is released under the [MIT License](http://opensource.org/licenses/MIT)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [0.2.1]

* Update docstring to mention constant time comparisons. No code level changes have happened.

## [0.2.0]

* Ring middleware for IP whitelisting


[Unreleased]: https://github.com/danielcompton/ring-ip-whitelisting/compare/0.1.0...HEAD
[0.1.1]: https://github.com/danielcompton/ring-ip-whitelisting/compare/0.1.0...0.1.1
[0.1.1]: https://github.com/danielcompton/ring-ip-whitelisting/compare/0.1.0...0.1.1


test/ring/middleware/ip_whitelist_test.clj
(ns ring.middleware.ip-whitelist-test
  (:require [clojure.test :refer :all]
            [ring.middleware.ip-whitelist :as ip]
            [criterium.core :as bench]
            [ring.mock.request :as mock]))

;; Utils

(defn success? [req]
  (= 200 (:status req)))

;; Test internals

(deftest cidr-range-test
  (are [start end cidr] (= [start end] (ip/cidr-range cidr))
    1 1 "0.0.0.1/32"
    0 63 "0.0.0.1/26"
    0 63 "0.0.0.1/26"
    0 63 "0.0.0.62/26"
    2130706432 2130706433 "127.0.0.1/31"
    256 256 "0.0.1.0/32"
    1 1 "::1/128"
    0 0 "::0/128"
    0 79228162514264337593543950335 "::1/32"
    256 256 "::ffff:0.0.1.0/128"
    ))

(deftest normalise-cidr-test
  (are [expected given] (= expected (ip/normalise-cidr given))
    "::1/128" "::1"
    "127.0.0.1/32" "127.0.0.1"
    "210.11.25.12/24" "210.11.25.12/24"
    "2001:0db8:85a3:0000:0000:8a2e:0370:7334/120" "2001:0db8:85a3:0000:0000:8a2e:0370:7334/120"))

(def whitelist ["201.202.161.33" "62.99.73.200/30" "177.10.156.0/21" "32.62.57.106/25"
                "::ffff:0.0.1.0/128" "2001:0db8:85a3:0000:0000:8a2e:0370:7334/120" "::2/127"])

(deftest in-whitelist-test
  (let [cidrs whitelist
        ip-set (ip/build-ip-set cidrs)]
    (are [in? ip] (= in? (ip/ip-in-ip-set? ip-set ip))
      true "201.202.161.33"
      false "201.202.161.34"
      true "62.99.73.200"
      true "62.99.73.202"
      true "62.99.73.203"
      false "62.99.73.204"
      true "177.10.156.0"
      true "177.10.159.255"
      false "177.10.160.0"
      false "0.0.0.0"
      false "255.255.255.255"
      false "::0"
      false "::1"
      true "::2"
      true "::3"
      true "0.0.1.0"
      true "2001:0db8:85a3:0000:0000:8a2e:0370:7334"
      true "2001:0db8:85a3:0000:0000:8a2e:0370:7385"
      )))


;; Test middleware

(def response {:status 200, :headers {} :body "Hi!"})

(deftest ip-whitelisting-test
  (let [handler (ip/wrap-ip-whitelist
                  (constantly response)
                  {:cidrs whitelist})]
    (are [success ip] (= (-> (mock/request :get "/")
                             (assoc :remote-addr ip)
                             (handler)
                             (success?))
                         success)
      true "201.202.161.33"
      false "201.202.161.34"
      true "62.99.73.200"
      true "62.99.73.202"
      true "62.99.73.203"
      false "62.99.73.204"
      true "177.10.156.0"
      true "177.10.159.255"
      false "177.10.160.0"
      false "0.0.0.0"
      false "255.255.255.255"
      false "::0"
      false "::1"
      true "::2"
      true "::3"
      true "0.0.1.0"
      true "2001:0db8:85a3:0000:0000:8a2e:0370:7334"
      true "2001:0db8:85a3:0000:0000:8a2e:0370:7385"
      )
    (is (= "<h1>Not authorized</h1>"
           (-> (mock/request :get "/")
               (assoc :remote-addr "0.0.0.0")
               (handler)
               :body)))))

(deftest ip-fn-test
  (let [handler (ip/wrap-ip-whitelist
                  (constantly response)
                  {:cidrs whitelist
                   :ip-fn :x-forwarded-for})]
    (are [success req] (= (-> req
                              (handler)
                              (success?))
                          success)
      true (assoc (mock/request :get "/")
             :x-forwarded-for "201.202.161.33")
      true (assoc (mock/request :get "/")
             :x-forwarded-for "201.202.161.33"
             :remote-addr "1.1.1.1")
      false (assoc (mock/request :get "/")
              :x-forwarded-for "2.2.2.2"
              :remote-addr "201.202.161.33")
      false (assoc (mock/request :get "/")
              :remote-addr "201.202.161.33"))))

(deftest error-response-test
  (let [handler (ip/wrap-ip-whitelist
                  (constantly response)
                  {:cidrs          whitelist
                   :error-response (constantly {:status 404 :headers {} :body "Not found"})})]
    (is (= 404 (-> (mock/request :post "/secret")
                   (assoc :remote-addr "8.8.4.4")
                   (handler)
                   :status)))
    (is (= 200 (-> (mock/request :post "/secret")
                   (assoc :remote-addr "177.10.156.0")
                   (handler)
                   :status)))))

(deftest allow-access?-test
  (let [handler (ip/wrap-ip-whitelist
                  (constantly response)
                  {:cidrs         whitelist
                   :allow-access? (fn [request]
                                    (= (:query-string request) "password=knockknock"))})]
    (is (= 200 (-> (mock/request :get "/admin?password=knockknock")
                   (assoc :remote-addr "8.8.4.4")
                   (handler)
                   :status)))
    (is (= 403 (-> (mock/request :get "/admin")
                   (assoc :remote-addr "8.8.4.4")
                   (handler)
                   :status)))))

(deftest cidr-validation
  (is (thrown? AssertionError
               (ip/wrap-ip-whitelist
                 (constantly response)
                 {})))
  (is (thrown? AssertionError
               (ip/wrap-ip-whitelist
                 (constantly response)
                 {:cidrs "100.200.100.200"}))))

(deftest updating-cidr-atom-test
  (let [cidr-list (atom [])
        handler (ip/wrap-ip-whitelist
                  (constantly response)
                  {:cidrs         cidr-list})]
    (is (= 403 (-> (mock/request :get "/admin")
                   (assoc :remote-addr "8.8.4.4")
                   (handler)
                   :status)))
    (swap! cidr-list conj "8.8.4.4")
    (is (= 200 (-> (mock/request :get "/admin")
                   (assoc :remote-addr "8.8.4.4")
                   (handler)
                   :status)))
    (reset! cidr-list [])
    (is (= 403 (-> (mock/request :get "/admin")
                   (assoc :remote-addr "8.8.4.4")
                   (handler)
                   :status)))))

;; Benchmarks

;(defn benchmark []
;  (let [cidrs ["201.202.161.33" "62.99.73.200/30" "177.10.156.0/21" "32.62.57.106/25"]
;        ip-set (ip/build-ip-set cidrs)]
;    (bench/bench (ip/ip-in-ip-set? ip-set "177.10.160.0"))))
;;; 53 ns
;
;;; This is an alternate implementation where we loop over each subnet and check if the IP is in it.
;;; Roughly 20x slower. Trades off smaller space for slower execution.
;;; Not needed at the moment, may be useful in the future.
;
;(defn build-ip-subnet-set [cidrs]
;  (into #{} (map ip/cidr->subnet-util) cidrs))
;
;(defn in-subnet-whitelist [subnet-set ^String ip]
;  (reduce (fn [in-whitelist? ^SubnetUtils subnet]
;            (if (.isInRange (.getInfo subnet) ip)
;              (reduced true)
;              false))
;          false
;          subnet-set))
;
;(defn benchmark-2 []
;  (let [cidrs ["201.202.161.33" "62.99.73.200/30" "177.10.156.0/21" "32.62.57.106/25"]
;        ip-set (build-ip-subnet-set cidrs)]
;    (bench/bench (in-subnet-whitelist ip-set "177.10.160.0"))))
;;; 1100 ns


test/ring/middleware/ip_whitelisting_test.clj
(ns ring.middleware.ip-whitelisting-test
  (:require [clojure.test :refer :all]
            [ring.middleware.ip-whitelisting :as ip]
            [criterium.core :as bench]
            [ring.mock.request :as mock]))

;; Utils

(defn success? [req]
  (= 200 (:status req)))

;; Test internals

(deftest cidr-range-test
  (are [start end cidr] (= [start end] (ip/cidr-range cidr))
    1 1 "0.0.0.1/32"
    0 63 "0.0.0.1/26"
    0 63 "0.0.0.1/26"
    0 63 "0.0.0.62/26"
    2130706432 2130706433 "127.0.0.1/31"
    256 256 "0.0.1.0/32"
    1 1 "::1/128"
    0 0 "::0/128"
    0 79228162514264337593543950335 "::1/32"
    256 256 "::ffff:0.0.1.0/128"
    ))

(deftest normalise-cidr-test
  (are [expected given] (= expected (ip/normalise-cidr given))
    "::1/128" "::1"
    "127.0.0.1/32" "127.0.0.1"
    "210.11.25.12/24" "210.11.25.12/24"
    "2001:0db8:85a3:0000:0000:8a2e:0370:7334/120" "2001:0db8:85a3:0000:0000:8a2e:0370:7334/120"))

(def whitelist ["201.202.161.33" "62.99.73.200/30" "177.10.156.0/21" "32.62.57.106/25"
                "::ffff:0.0.1.0/128" "2001:0db8:85a3:0000:0000:8a2e:0370:7334/120" "::2/127"])

(deftest in-whitelist-test
  (let [cidrs whitelist
        ip-set (ip/build-ip-set cidrs)]
    (are [in? ip] (= in? (ip/ip-in-ip-set? ip-set ip))
      true "201.202.161.33"
      false "201.202.161.34"
      true "62.99.73.200"
      true "62.99.73.202"
      true "62.99.73.203"
      false "62.99.73.204"
      true "177.10.156.0"
      true "177.10.159.255"
      false "177.10.160.0"
      false "0.0.0.0"
      false "255.255.255.255"
      false "::0"
      false "::1"
      true "::2"
      true "::3"
      true "0.0.1.0"
      true "2001:0db8:85a3:0000:0000:8a2e:0370:7334"
      true "2001:0db8:85a3:0000:0000:8a2e:0370:7385"
      )))


;; Test middleware

(def response {:status 200, :headers {} :body "Hi!"})

(deftest ip-whitelisting-test
  (let [handler (ip/wrap-ip-whitelisting
                  (constantly response)
                  {:cidrs whitelist})]
    (are [success ip] (= (-> (mock/request :get "/")
                             (assoc :remote-addr ip)
                             (handler)
                             (success?))
                         success)
      true "201.202.161.33"
      false "201.202.161.34"
      true "62.99.73.200"
      true "62.99.73.202"
      true "62.99.73.203"
      false "62.99.73.204"
      true "177.10.156.0"
      true "177.10.159.255"
      false "177.10.160.0"
      false "0.0.0.0"
      false "255.255.255.255"
      false "::0"
      false "::1"
      true "::2"
      true "::3"
      true "0.0.1.0"
      true "2001:0db8:85a3:0000:0000:8a2e:0370:7334"
      true "2001:0db8:85a3:0000:0000:8a2e:0370:7385"
      )
    (is (= "<h1>Not authorized</h1>"
           (-> (mock/request :get "/")
               (assoc :remote-addr "0.0.0.0")
               (handler)
               :body)))))

(deftest ip-fn-test
  (let [handler (ip/wrap-ip-whitelisting
                  (constantly response)
                  {:cidrs whitelist
                   :ip-fn :x-forwarded-for})]
    (are [success req] (= (-> req
                              (handler)
                              (success?))
                          success)
      true (assoc (mock/request :get "/")
             :x-forwarded-for "201.202.161.33")
      true (assoc (mock/request :get "/")
             :x-forwarded-for "201.202.161.33"
             :remote-addr "1.1.1.1")
      false (assoc (mock/request :get "/")
              :x-forwarded-for "2.2.2.2"
              :remote-addr "201.202.161.33")
      false (assoc (mock/request :get "/")
              :remote-addr "201.202.161.33"))))

(deftest error-response-test
  (let [handler (ip/wrap-ip-whitelisting
                  (constantly response)
                  {:cidrs          whitelist
                   :error-response (constantly {:status 404 :headers {} :body "Not found"})})]
    (is (= 404 (-> (mock/request :post "/secret")
                   (assoc :remote-addr "8.8.4.4")
                   (handler)
                   :status)))
    (is (= 200 (-> (mock/request :post "/secret")
                   (assoc :remote-addr "177.10.156.0")
                   (handler)
                   :status)))))

(deftest allow-access?-test
  (let [handler (ip/wrap-ip-whitelisting
                  (constantly response)
                  {:cidrs         whitelist
                   :allow-access? (fn [request]
                                    (= (:query-string request) "password=knockknock"))})]
    (is (= 200 (-> (mock/request :get "/admin?password=knockknock")
                   (assoc :remote-addr "8.8.4.4")
                   (handler)
                   :status)))
    (is (= 403 (-> (mock/request :get "/admin")
                   (assoc :remote-addr "8.8.4.4")
                   (handler)
                   :status)))))

(deftest cidr-validation
  (is (thrown? AssertionError
               (ip/wrap-ip-whitelisting
                 (constantly response)
                 {})))
  (is (thrown? AssertionError
               (ip/wrap-ip-whitelisting
                 (constantly response)
                 {:cidrs "100.200.100.200"}))))

(deftest updating-cidr-atom-test
  (let [cidr-list (atom [])
        handler (ip/wrap-ip-whitelisting
                  (constantly response)
                  {:cidrs         cidr-list})]
    (is (= 403 (-> (mock/request :get "/admin")
                   (assoc :remote-addr "8.8.4.4")
                   (handler)
                   :status)))
    (swap! cidr-list conj "8.8.4.4")
    (is (= 200 (-> (mock/request :get "/admin")
                   (assoc :remote-addr "8.8.4.4")
                   (handler)
                   :status)))
    (reset! cidr-list [])
    (is (= 403 (-> (mock/request :get "/admin")
                   (assoc :remote-addr "8.8.4.4")
                   (handler)
                   :status)))))

;; Benchmarks

;(defn benchmark []
;  (let [cidrs ["201.202.161.33" "62.99.73.200/30" "177.10.156.0/21" "32.62.57.106/25"]
;        ip-set (ip/build-ip-set cidrs)]
;    (bench/bench (ip/ip-in-ip-set? ip-set "177.10.160.0"))))
;;; 53 ns
;
;;; This is an alternate implementation where we loop over each subnet and check if the IP is in it.
;;; Roughly 20x slower. Trades off smaller space for slower execution.
;;; Not needed at the moment, may be useful in the future.
;
;(defn build-ip-subnet-set [cidrs]
;  (into #{} (map ip/cidr->subnet-util) cidrs))
;
;(defn in-subnet-whitelist [subnet-set ^String ip]
;  (reduce (fn [in-whitelist? ^SubnetUtils subnet]
;            (if (.isInRange (.getInfo subnet) ip)
;              (reduced true)
;              false))
;          false
;          subnet-set))
;
;(defn benchmark-2 []
;  (let [cidrs ["201.202.161.33" "62.99.73.200/30" "177.10.156.0/21" "32.62.57.106/25"]
;        ip-set (build-ip-subnet-set cidrs)]
;    (bench/bench (in-subnet-whitelist ip-set "177.10.160.0"))))
;;; 1100 ns


src/ring/middleware/ip_whitelisting.clj
(ns ring.middleware.ip-whitelisting
  (:refer-clojure :exclude [biginteger])
  (:require [clojure.string :as str])
  (:import [clojure.lang IAtom]
           [java.net InetAddress]
           [java.nio ByteBuffer]))

(defn ^BigInteger inet->value
  [^InetAddress inet-address]
  (let [buffer (ByteBuffer/wrap (.getAddress inet-address))]
    (BigInteger. 1 (.array buffer))))

(defn ip-value
  "Returns the value of an IP (v4 or v6) address string"
  [ip-address]
  (inet->value (InetAddress/getByName ip-address)))

(defn normalise-cidr [^String cidr-or-ip]
  (let [])
  (if (str/index-of cidr-or-ip "/") ;; If there is no slash in the cidr, treat it as a /32.
    cidr-or-ip
    (->> (.. (InetAddress/getByName cidr-or-ip) getAddress)
         count ;; count number of bytes in address
         (* 8) ;; * 8 to get address size
         (str cidr-or-ip "/")))) ;; append address size to IP address to only match the single IP address

(defn cidr-range
  "Takes a cidr string and returns the start and end IP value in the range."
  [^String cidr]
  ;; Logic partially based on https://github.com/edazdarevic/CIDRUtils/blob/master/CIDRUtils.java
  (let [index (.indexOf cidr "/")
        _ (assert (pos? index) "CIDR must include /<n>")
        [address network] (str/split cidr #"/" 2)
        inet-address (InetAddress/getByName address)
        prefix-length (Integer/parseInt network)
        address-length (count (.getAddress inet-address)) ;; 4 (IPv4) or 16 (IPv6)
        address-mask (.. (BigInteger. 1 (byte-array address-length (repeat address-length -1))) ;; Fill array with all bits set to 1
                         not
                         (shiftRight prefix-length))
        ip-val (inet->value inet-address)
        start-val (.and ip-val address-mask)
        end-val (.add start-val (.not address-mask))]
    [start-val end-val]))

(defn build-ip-set
  "Returns a set which contains every IP address that is in the range of one of the CIDRs.

  * cidrs: a sequence of cidr strings or IP addresses."
  [cidrs]
  (into #{}
        (comp (map normalise-cidr)
              (map cidr-range)
              (mapcat (fn [[start end]]
                        (range start (inc end)))))
        cidrs))

(defn ip-in-ip-set? [ip-set ip]
  (contains? ip-set (ip-value ip)))

(defn- access-denied [body]
  {:status  403
   :headers {"Content-Type" "text/html"}
   :body    body})

(defn wrap-ip-whitelisting
  "Middleware that secures routes by IP whitelist. Checks the IP address of every request
  and denies access if requester is not in the whitelist.

  IP whitelisting can provide additional security, but can be spoofed or otherwise defeated.
  It shouldn't be the only line of defense for sensitive resources.

  * cidrs: a whitelist of CIDR strings or IP addresses. IP addresses are treated as a CIDR for
    the single address (e.g. a /32 for an IPv4 address). If the list is inside an atom
    then the whitelist will be updated when the atom changes.
  * ip-fn: function to extract the IP address. Defaults to `:remote-addr`
  * error-response: Response to return if request is not authorised
  * allow-access?: An additional function that can be provided if IP whitelisting fails but
    more checking is needed before denying access. Takes a request argument, returns true if
    the request should be allowed, else false."
  [handler {:keys [cidrs ip-fn error-response allow-access?]
            :as   options
            :or   {ip-fn          :remote-addr
                   error-response #(access-denied "<h1>Not authorized</h1>")
                   allow-access?  (constantly false)}}]
  (let [cidrs-atom? (instance? IAtom cidrs)
        dereffed-cidrs (if cidrs-atom? @cidrs cidrs)
        _ (assert (sequential? dereffed-cidrs) "cidrs must be a sequential collection")
        _ (assert (every? string? dereffed-cidrs) "every element in cidrs must be a string")
        ip-whitelist (atom (build-ip-set dereffed-cidrs))]
    (when cidrs-atom?
      (add-watch cidrs :watcher (fn [_ _ _ new-state]
                                  (reset! ip-whitelist (build-ip-set new-state)))))
    (fn [request]
      (if (or (ip-in-ip-set? @ip-whitelist (ip-fn request))
              (allow-access? request))
        (handler request)
        (error-response)))))


src/ring/middleware/ip_whitelist.clj
(ns ring.middleware.ip-whitelist
  (:refer-clojure :exclude [biginteger])
  (:require [clojure.string :as str])
  (:import [clojure.lang IAtom]
           [java.net InetAddress]
           [java.nio ByteBuffer]))

(defn ^BigInteger inet->value
  [^InetAddress inet-address]
  (let [buffer (ByteBuffer/wrap (.getAddress inet-address))]
    (BigInteger. 1 (.array buffer))))

(defn ip-value
  "Returns the value of an IP (v4 or v6) address string"
  [ip-address]
  (inet->value (InetAddress/getByName ip-address)))

(defn normalise-cidr [^String cidr-or-ip]
  (let [])
  (if (str/index-of cidr-or-ip "/") ;; If there is no slash in the cidr, treat it as a CIDR for a single IP.
    cidr-or-ip
    (->> (.. (InetAddress/getByName cidr-or-ip) getAddress)
         count ;; count number of bytes in address
         (* 8) ;; * 8 to get address size
         (str cidr-or-ip "/")))) ;; append address size to IP address to only match the single IP address

(defn cidr-range
  "Takes a cidr string and returns the start and end IP value in the range."
  [^String cidr]
  ;; Logic partially based on https://github.com/edazdarevic/CIDRUtils/blob/master/CIDRUtils.java
  (let [index (.indexOf cidr "/")
        _ (assert (pos? index) "CIDR must include /<n>")
        [address network] (str/split cidr #"/" 2)
        inet-address (InetAddress/getByName address)
        prefix-length (Integer/parseInt network)
        address-length (count (.getAddress inet-address)) ;; 4 (IPv4) or 16 (IPv6)
        address-mask (.. (BigInteger. 1 (byte-array address-length (repeat address-length -1))) ;; Fill array with all bits set to 1
                         not
                         (shiftRight prefix-length))
        ip-val (inet->value inet-address)
        start-val (.and ip-val address-mask)
        end-val (.add start-val (.not address-mask))]
    [start-val end-val]))

(defn build-ip-set
  "Returns a set which contains every IP address that is in the range of one of the CIDRs.

  * cidrs: a sequence of cidr strings or IP addresses."
  [cidrs]
  (into #{}
        (comp (map normalise-cidr)
              (map cidr-range)
              (mapcat (fn [[start end]]
                        (range start (inc end)))))
        cidrs))

(defn ip-in-ip-set? [ip-set ip]
  (contains? ip-set (ip-value ip)))

(defn- access-denied [body]
  {:status  403
   :headers {"Content-Type" "text/html"}
   :body    body})

(defn wrap-ip-whitelist
  "Middleware that secures routes by IP whitelist. Checks the IP address of every request
  and denies access if requester is not in the whitelist.

  IP whitelisting can provide additional security, but can be spoofed or otherwise defeated.
  It shouldn't be the only line of defense for sensitive resources.

  * cidrs: a whitelist of CIDR strings or IP addresses. IP addresses are treated as a CIDR for
    the single address (e.g. a /32 for an IPv4 address). If the list is inside an atom
    then the whitelist will be updated when the atom changes.
  * ip-fn: function to extract the IP address. Defaults to `:remote-addr`
  * error-response: Response to return if request is not authorised
  * allow-access?: An additional function that can be provided if IP whitelisting fails but
    more checking is needed before denying access. Takes a request argument, returns true if
    the request should be allowed, else false. Note, if you're checking for a secret request parameter
    you probably shouldn't be, but if you still want to, make sure you do the comparison with a
    constant time equals, e.g. https://github.com/weavejester/crypto-equality"
  [handler {:keys [cidrs ip-fn error-response allow-access?]
            :as   options
            :or   {ip-fn          :remote-addr
                   error-response (fn [req] (access-denied "<h1>Not authorized</h1>"))
                   allow-access?  (constantly false)}}]
  (let [cidrs-atom? (instance? IAtom cidrs)
        dereffed-cidrs (if cidrs-atom? @cidrs cidrs)
        _ (assert (sequential? dereffed-cidrs) "cidrs must be a sequential collection")
        _ (assert (every? string? dereffed-cidrs) "every element in cidrs must be a string")
        ip-whitelist (atom (build-ip-set dereffed-cidrs))]
    (when cidrs-atom?
      (add-watch cidrs :watcher (fn [_ _ _ new-state]
                                  (reset! ip-whitelist (build-ip-set new-state)))))
    (fn [request]
      (if (or (ip-in-ip-set? @ip-whitelist (ip-fn request))
              (allow-access? request))
        (handler request)
        (error-response request)))))
